drv/gimlet-hf-server/src/main.rs

```diff
@@ -376,36 +698,95 @@ impl idl::InOrderHostFlashImpl for ServerImpl {
-        if status & 1 == 0 {
-            // ooh we're done
-            break;
-        }
-        if let Some(ticks) = sleep_between_polls {
-            hl::sleep_for(ticks);
+    /// Writes the given persistent data to host flash
+    fn write_persistent_data(
+        &mut self,
+        _: &RecvMessage,
+        startup_options: u64,
+        dev_select: HfDevSelect,
+    ) -> Result<(), RequestError<HfError>> {
+        let data = HfPersistentData {
+            startup_options,
+            dev_select,
+        };
+        self.check_muxed_to_sp()?;
+        if self.dev_select_pin.is_some() {
+            let prev_slot = self.dev_state;
+
+            // At this point, all of the ways that `set_dev` could fail have
+            // been checked: we already called both `check_muxed_to_sp` and that
+            // `dev_select_pin` is `Some(...)`, so we can unwrap returns from
+            // `self.set_dev(...)`.
+            self.set_dev(!prev_slot).unwrap();
+            let (a_data, a_next) = self.persistent_data_scan();
+
+            self.set_dev(prev_slot).unwrap();
+            let (b_data, b_next) = self.persistent_data_scan();
+
+            let prev_epoch = match (a_data, b_data) {
+                (Some(a), Some(b)) => a.epoch.max(b.epoch),
+                (Some(a), None) => a.epoch,
+                (None, Some(b)) => b.epoch,
+                (None, None) => 0,
+            };
+
+            // Early exit if the previous persistent data matches
+            let prev_raw = RawPersistentData::new(data, prev_epoch);
+            if a_data == b_data && a_data == Some(prev_raw) {
+                return Ok(());
+            }
+
+            let epoch = prev_epoch + 1;
+            let raw = RawPersistentData::new(data, epoch);
+
+            // Write the persistent data to the currently inactive flash.
+            self.set_dev(!prev_slot).unwrap();
+            let out_a = self.write_raw_persistent_data_to_addr(a_next, &raw);
+
+            // Swap back to the currently selected flash
+            self.set_dev(prev_slot).unwrap();
+
+            // Now that we've restored the current active flash, check whether
+            // we should propagate errors.
+            out_a?;
+
+            // Write the persistent data to the currently active flash
+            self.write_raw_persistent_data_to_addr(b_next, &raw)?;
+        } else {
+            // Single-flash case is less complicated, because we don't have to
+            // track and maintain the active device.
+            let (prev_data, next) = self.persistent_data_scan();
+            let prev_epoch = match prev_data {
+                Some(a) => a.epoch,
+                None => 0,
+            };
+
+            // Early exit if the previous persistent data matches
+            let prev_raw = RawPersistentData::new(data, prev_epoch);
+            if prev_data == Some(prev_raw) {
+                return Ok(());
+            }
+
+            let epoch = prev_epoch + 1;
```

Comment:
It's a `u64`, rather than a `u32`, so it _definitely_ will not roll over unless someone manually reprograms the flash to have data with an artificially high epoch.  I added an overflow check in 53d6b530c914c8f0f6c862f4305f75ab89475d0e which will return an error on that acse.